components/ftx/browser/ftx_service.cc

```diff
@@ -0,0 +1,398 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/ftx/browser/ftx_service.h"
+
+#include <string>
+#include <utility>
+
+#include "base/base64.h"
+#include "base/json/json_writer.h"
+#include "base/task/post_task.h"
+#include "base/task/thread_pool.h"
+#include "base/task_runner_util.h"
+#include "brave/components/ftx/browser/ftx_json_parser.h"
+#include "brave/components/ftx/common/pref_names.h"
+#include "brave/components/ntp_widget_utils/browser/ntp_widget_utils_oauth.h"
+#include "components/os_crypt/os_crypt.h"
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/storage_partition.h"
+#include "net/base/escape.h"
+#include "net/base/load_flags.h"
+#include "net/base/url_util.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "services/network/public/mojom/fetch_api.mojom-shared.h"
+
+namespace {
+
+const char api_host[] = "ftx.com";
+const char oauth_callback[] = "com.brave.ftx://authorization";
+const unsigned int kRetriesCountOnNetworkChange = 1;
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("ftx_service", R"(
+      semantics {
+        sender: "FTX Service"
+        description:
+          "This service is used to communicate with FTX "
+          "on behalf of the user interacting with the FTX widget."
+        trigger:
+          "Triggered by using the FTX widget."
+        data:
+          "Account balance for the widget."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        setting:
+          "You can enable or disable this feature on the new tab page."
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+GURL GetURLWithPath(const std::string& host, const std::string& path) {
+  return GURL(std::string(url::kHttpsScheme) + "://" + host).Resolve(path);
+}
+
+void BuildFormEncoding(const std::string& key,
+                       const std::string& value,
+                       std::string* out) {
+  if (!out->empty())
+    out->append("&");
+  out->append(key + "=" + net::EscapeUrlEncodedData(value, true));
+}
+
+}  // namespace
+
+FTXService::FTXService(content::BrowserContext* context)
```

Comment:
passing in BrowserContext should be avoided because it makes this impossible to use for ios and is generally considered an anti-pattern. It is better to pass in the url_loader_factory and user prefs to avoid any content dependencies

Reply:
follow-up https://github.com/brave/brave-browser/issues/15851